crates/macro-support/src/parser.rs

```diff
@@ -911,24 +912,24 @@ fn function_from_decl(
 
     let syn::Signature { inputs, output, .. } = sig;
 
-    let replace_self = |t: syn::Type| {
-        let self_ty = match self_ty {
-            Some(i) => i,
-            None => return t,
-        };
-        let path = match get_ty(&t) {
-            syn::Type::Path(syn::TypePath { qself: None, path }) => path.clone(),
-            other => return other.clone(),
-        };
-        let new_path = if path.segments.len() == 1 && path.segments[0].ident == "Self" {
-            self_ty.clone().into()
-        } else {
-            path
-        };
-        syn::Type::Path(syn::TypePath {
-            qself: None,
-            path: new_path,
-        })
+    // A helper function to replace `Self` in the function signature of methods
+    // The following comment explains why this is necessary:
+    // https://github.com/rustwasm/wasm-bindgen/issues/3105#issuecomment-1275160744
+    let replace_self = |mut t: syn::Type| {
+        if let Some(self_ty) = self_ty {
+            struct SelfReplace(Ident);
+            impl VisitMut for SelfReplace {
+                fn visit_ident_mut(&mut self, i: &mut proc_macro2::Ident) {
+                    if i == "Self" {
+                        *i = self.0.clone();
+                    }
+                }
+            }
+
+            let mut replace = SelfReplace(self_ty.clone());
+            replace.visit_type_mut(&mut t);
+        }
+        t
```

Comment:
Apologies, I wasn't being specific enough.
This is the part I think that needs to be explained.

Whats happening here?
How is this used to replace all `Self` regardless of depth? (is that the case?)
Why is `VisitMut` necessary?

This might need more comments in the surrounding code as well.
You didn't write all the surrounding code, I'm sorry this is being pushed onto you!

Reply:
Ah, sorry about that. I did **a lot** of stuff with ASTs before, so using visitors is almost second nature to me at this point.

I'll add some more comments to answer your questions.

> You didn't write all the surrounding code, I'm sorry this is being pushed onto you!

No worries. I'm currently working on another PR that changes some code in this area, so I can add a few comments to explain the big picture and how all the pieces for together.

Reply:
Okay, I added a few comments **for now**.

I actually wrote a few more, but discovered some minor bugs while doing that. So I'll fix those bugs and explain the fixed code in a follow-up PR.